### PR TITLE
Support data attributes

### LINKED ISSUE #43: Allow `data-` attrs for valid HTML5
No page that uses Twine is valid HTML because of the custom attributes. I'm not suggesting killing these unprefixed versions, but it would be great if we could alias all of Twine’s attributes (e.g. `bind`, `define`) to data attributes (e.g. `data-bind`, `data-define`). That way any users that care about having valid markup can use the `data-` ones, and those that don’t can continue using the unprefixed versions.

### PR DESCRIPTION
**Pull request by nwtn**
Fixes #43.

### CODE REVIEW

File: lib/assets/javascripts/twine.js.coffee [hunk 1]
```diff
@@ -326,5 +336,8 @@ setupEventBinding = (eventName) ->
     return teardown: ->
       $(node).off eventName, onEventHandler
 
+  Twine.bindingTypes["data-bind-event-#{eventName}"] = Twine.bindingTypes["bind-event-#{eventName}"]
```
Comment (DrewMartin) [inline lib/assets/javascripts/twine.js.coffee]: Probably cleaner to have a line at the end that does something like
```coffeescript
for key in Object.keys(Twine.bindingTypes)
  Twine.bindingTypes["data-#{key}"] = Twine.bindingTypes[key]
```

And then have a test that ensures that for every `bind-whatever

**Comment by qq99** [comment]
cc @Shopify/tnt

File: dist/twine.js [hunk 2]
```diff
@@ -307,8 +307,8 @@ wrapFunctionString = function(code, args, node) {
   } else {
     try {
       return new Function(args, "with($context) { return " + code + " }");
-    } catch (_error) {
-      e = _error;
+    } catch (error) {
+      e = error;
```
Comment (qq99) [inline dist/twine.js]: do we ever use `e` or `error`?

File: lib/assets/javascripts/twine.js.coffee (hunk 1 above)
Comment (qq99) [inline lib/assets/javascripts/twine.js.coffee]: :+1:

File: dist/twine.js (hunk 2 above)
Comment (celsodantas) [inline dist/twine.js]: I wonder why this is being renamed.

**Comment by DrewMartin** [comment]
Oh you know what? This will lead to subtle bugs. We kind of depend on the order of the keys in `Twine.bindingTypes` so that when we go over the bindings on a node (https://github.com/Shopify/twine/blob/data-attributes/lib/assets/javascripts/twine.js.coffee#L48-L51) we do them in a specific order. If someone were to use a mix of data and non-data attributes, say `<input type="text" bind-event-input="onInput()" data-bind="val">` it would not work like you expect because the input event triggering `onInput()` would be fired before the input event that changes `val` so inside of the `onInput` method it would have the old value when you look at `@val`. #31 had the same issue (fixed in #33)

File: dist/twine.js (hunk 2 above)
Comment (pushrax) [inline dist/twine.js]: CoffeeScript doing CoffeeScript things.

**Comment by DrewMartin** [comment]
Alternative approach: change line 48 to
`for type, binding of Twine.bindingTypes when definition = (node.getAttribute(type) || node.getAttribute("data-#{type}"))`

**Comment by pushrax** [comment]
Approach @DrewMartin gave is probably the best one, despite the associated performance cost. It might be good to have this feature enabled by configuration.

**Comment by qq99** [comment]
Good catch @DrewMartin

**Comment by nwtn** [comment]
Great feedback, thanks. I've updated with the suggested change from @DrewMartin.

**Comment by nwtn** [comment]
ping @DrewMartin @pushrax @qq99

**Comment by DrewMartin** [comment]
:+1:

**Comment by qq99** [comment]
LGTM

**Comment by nwtn** [comment]
cool cool cool. is there anything that needs to be done besides merging and following the [release steps](https://github.com/Shopify/twine#releasing)? e.g. shipit?

**Comment by qq99** [comment]
iirc, that's all

I thought we were listed on bower at one point (https://github.com/Shopify/twine/blob/master/bower.json) but I am not seeing us on the registry

**Pull request merged**